aten/src/ATen/cuda/CUDAGraph.cpp

```diff
@@ -315,4 +338,261 @@ CUDAGraph::~CUDAGraph() {
+    if (ptr >= begin && ptr < end) {
+      size_t offset = (size_t)((char*)ptr - (char*)begin);
+      return std::make_tuple(i, offset);
+    }
+  }
+  return std::nullopt;
+};
+
+std::vector<TrackedAllocation> combine_ranges(const std::vector<TrackedAllocation>& input, size_t num_dynamic_args) {
+    std::vector<TrackedAllocation> ranges = input;
+    // combine according to overlap:
+    std::sort(ranges.begin(), ranges.end(), [](const TrackedAllocation& a, const TrackedAllocation& b) {
+      return a.ptr < b.ptr;
+    });
+    std::vector<TrackedAllocation> result;
+    for (const TrackedAllocation& range : ranges) {
+      if (result.empty() || range.ptr >= result.back().ptr + result.back().size) { // ">=" because ranges are exclusive on upper end
+        result.push_back(range); // No overlap, add new range
+      } else {
+        TrackedAllocation& last = result.back(); // Overlap found, extend the previous range
+        TORCH_CHECK(last.alloc_idx >= num_dynamic_args && range.alloc_idx >= num_dynamic_args, "Cannot combine argument allocations");
+        char* last_end = last.ptr + last.size;
+        char* range_end = range.ptr + range.size;
+        last.size = std::max(last_end, range_end) - last.ptr;
+      }
+    }
+    // sort by original alloc_idx to bring the original dynamic args back to the beginning, in original order
+    // (since the dynamic args were allocated from a private pool, they should always be in ascending ptr order anyway? but let's just be safe)
+    std::sort(result.begin(), result.end(), [](const TrackedAllocation& a, const TrackedAllocation& b) {
+      return a.alloc_idx < b.alloc_idx;
+    });
+    return result;
+}
+
+void CUDAGraph::introspect_dynamic_graph() {
+  allocations_ = combine_ranges(allocations_, num_dynamic_args_);
+  size_t num_nodes;
+  AT_CUDA_CHECK(cudaGraphGetNodes(graph_, nullptr, &num_nodes));
+  std::cout << "number of nodes captured " << num_nodes << std::endl;
+  std::vector<cudaGraphNode_t> nodes(num_nodes);
+  AT_CUDA_CHECK(cudaGraphGetNodes(graph_, nodes.data(), &num_nodes));
+
+  for (size_t i = 0; i < num_nodes; ++i) {
+    cudaGraphNode_t node = nodes[i];
+
+    cudaGraphNodeType type;
+    AT_CUDA_CHECK(cudaGraphNodeGetType(node, &type));
+
+    if (type == cudaGraphNodeTypeKernel) {
+      cudaKernelNodeParams nodeParams;
+      AT_CUDA_CHECK(cudaGraphKernelNodeGetParams(node, &nodeParams));
+      cudaFunction_t func;
+      AT_CUDA_CHECK(cudaGetFuncBySymbol(&func, nodeParams.func));
+
+      const char* func_name;
+      globalContext().getNVRTC().cuFuncGetName(&func_name, func);
+
+      size_t param_index = 0;
+      size_t param_offset;
+      size_t param_size;
+      while (globalContext().getNVRTC().cuFuncGetParamInfo(
+                 func, param_index, &param_offset, &param_size) !=
+             CUDA_ERROR_INVALID_VALUE) {
+        char** arg1_speculative_pointer =
+            (char**)nodeParams.kernelParams[param_index];
```

Comment:
IIRC, kernelParams is empty if someone launched a cuda kernel using extra instead of kernelParams https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXEC.html#group__CUDA__EXEC_1gb8f3dc3031b40da29d5f9a7139e52e15

So we need to check if kernelParams is null. If it is, then the arguments are stored in the extra field of `nodeParams`. I can write the code for decoding the extra buffer if you want later. It's sort of an edge case anyway. I'm not aware of any pytorch code that uses cuLaunchKernel with the extra argument, but it's possible (e.g., in a custom op).